### DISCUSSION TITLE
Individualisation / Custom Music

### DISCUSSION THREAD

**Discussion by khoffinator** [beat-feet/beat-feet#80] [Ideas]
Hello Peter,

looking for a simple app being able using individual/personal songs I found your Beat-Game app...its pretty cool.

Maybe some background to my usecase: 
I am singing in a non-professional (non-commercial) choir in Germany, and due to pandemic rules, we are not able to do rehearsals, live sessions etc. To keep all members on board we try to keep them motivated by social media stuff, online rehearsals... AND also with a funny game playing to our own recordings.

So here some questions/findings regarding individualisation:
1) Custom Level: some songs I loaded into the specific folder unfortunately crashed the app. Some worked, but depending on the song, the level creator is doing some really tough stuff, if you have a full track (several voices, piano etc) --> maybe a regulator would be helpful to adjust the difficulty... "set obstacles per beat" ?!?

2) Customise Sprites: I tried to load new sprites to create a level with our choir look & logos (using Android Studio). I am also not quite sure, how the graphics are chosen and set into the level. In the first step, my idea was to manipulate the "sprites.png"... but trying to upload my new file, it seemed to be locked.. unfortunately, I am totally new to Android programming, too

So, hopefully you can help with this two topics? 
I can provide then the feedback from all choir members using your app ;-) 

Thanks in advance!

BR
Klaus

**Comment by pserwylo** [comment]
Thanks for taking such an interest in the app, facinating to hear about the way you plan on using it.

A few thoughts:
* If you are new to Android programming, it may be a little difficult to build a custom version with different sprites, but if you would like to continue, the best way to change them are to:
  * Find the original files in the `./texture-packer/sprites/` directory ([link](https://github.com/beat-feet/beat-feet/tree/master/texture-packer/sprites)).
  * From the root directory of the project, run `./gradlew :texture-packer:packTextures` which assembles the appropriate `sprites.png` along with the associated metadata to tell the game engine where to find each sprite in the resulting `.png`.

Keep in mind though, that even after doing that, you'll have to build and sign the resulting `.apk` file before sharing with others. There is plenty of tutorials for doign this on the internet though, and you can always ask here if you get this far.

I must say, I think the long term solution for this (and I stress long term because there are probably other more important issues to address first in terms of bugs or necessary features) would be to add the ability to easily mod the game for tasks like this, such that:
* The default list of levels is different.
* The graphics assets are different.
I would have originally thought this may be difficult to implement, however I've been reading the source of the [Unciv game for a bit lately and noticed that they way they implement mods is actually quite elegant and light-touch](https://github.com/yairm210/Unciv/wiki/Mods).

**Reply by khoffinator** [reply]
Hi, thanks for your reply and tipps for modding the sprites. I think for the beginning I will do the easiest way and just put out logo into the cloud sprites ;-)

Currently, I am working on getting used to Android Studio and reading turorials...still have some trouble to create an APK out of the code.
But in general: if I do not want to load the app into Google Play, I do not need that whole signing process, right?
For our purpose I think its ok just to share the APK via e.g. Dropbox with the members.

I understand your focus on the important topics. 
Let's see, what I can do :)

**Reply by khoffinator** [reply]
Hello Peter,
thanks again for your support. Meanwhile I figured out how to use the texture packer creating a new sprite sheet. That works well for me...nice to see own logos and graphics to own songs in the game ;-)

Maybe one more question: since we also have some members who are not so native with digital stuff, I would like to "pre-pack" the custom MP3 file into the APK. Means, when the game is installed, the requested folder "BeatChange" is created automatically and the custom file is placed inside. The game and the custom algorithm should just work as it is right now, so a change of the file is still possible.

Do you think, that can be done without big efforts?

**Reply by pserwylo** [reply]
Awesome, great to hear your making such good progress!

Regarding adding your song(s), I think we can do one better. The reason build-in songs packaged with the game load much faster than a custom MP3 does is because:
* The levels are generated offline on a PC.
* The generated level data is included in the `.apk` as a `.json` file.
* The songs are also downsampled to a lower bitrate to save size in the resulting `.apk` (but this may not be as helpful to you).

Therefore, I think you could offer a version where not only are your own songs pre-packaged, but they are shown in the main list of levels as the first playable levels (rather than hidden behind the "Custom" level), and they will load in a very quick time.

To do this:

**Place the song you want into `./songs/SONG_NAME.mp3`**

**Run `./gradlew :song-extract:processSongs`**

You should get output similar to the following:

```
[SongExtract] Processing the_haunted_mansion_the_exercise_room.mp3, writing to beat-game/song-extract/../android/assets/songs/data/the_haunted_mansion_the_exercise_room.json.
[WorldCache] Generating world from ../songs/original/the_haunted_mansion_the_exercise_room.mp3...
[WorldCache] Calculating FFT
[WorldCache] Extracting and smoothing features
[WorldCache] Extracting and smoothing height map
[WorldCache] Samples: 10245888 @ 44100Hz (duration: 232)
[WorldCache] Finished generating world
```
**(Optional - to save space but lower quality): Run `./songs/downsample.sh`**

If you don't run this, you'll need to manually ensure your song is copied to `./android/assets/songs/SONG_NAME.mp3`.

**Tell the game about your new level**

* Add a new entry to the `Levels` array in [Levels.kt](https://github.com/beat-feet/beat-feet/blob/master/core/src/com/serwylo/beatgame/levels/Level.kt#L14). 
* Note that the latest `master` I've linked to above takes a second argument that is an internationalised string key, which requires an entry in [messages.properties](https://github.com/beat-feet/beat-feet/blob/master/android/assets/i18n/messages.properties#L41) to define the name of the level. If you are based off an earlier version, you can directly hard code the name of the level here.
* Unlock it by default, so you don't need any achievements to play (the third argument to `Level` should be `Unlocked()`).
* The order of this list in `Levels.kt` is the order shown in the level select screen.

Let me know how that goes, or if it appears I've missed any steps once you get going, and I'll try to be a bit mroe detailed. Might add this to the readme somewhere too at some point.

**Reply by khoffinator** [reply]
Hey Peter, thanks for your explanation. Now that's way cooler, so I can create even more own levels :-) 

But actually I used the original level call, e.g. "LaundryRoom" and renamed mp3Name and labelID...not 100% clean, but it works.
I tried the analise function to find the root, where I can fully rename or create a new level, but I guess I used it wrong ;-)

**Reply by khoffinator** [reply]
Hi again...just want to let you know: forget my last message, I just made a little typing error and thought, it does not work.. 
Everything works fine now 👍 

Thanks again for you help!
